sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -210,6 +234,246 @@ public Row getCurrentRow() throws NoSuchElementException {
+      baseLabels.put(
+          MonitoringInfoConstants.Labels.INSTANCE_ID, session.getOptions().getInstanceId());
+      baseLabels.put(
+          MonitoringInfoConstants.Labels.TABLE_ID,
+          GcpResourceIdentifiers.bigtableTableID(
+              session.getOptions().getProjectId(),
+              session.getOptions().getInstanceId(),
+              source.getTableId().get()));
+      serviceCallMetric =
+          new ServiceCallMetric(MonitoringInfoConstants.Urns.API_REQUEST_COUNT, baseLabels);
+
+      //future = session.getDataClient().readRowsAsync(buildReadRowsRequest());
+      future = createFuture();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (buffer.size() <= miniBatchWaterMark && future == null && !lastFillComplete) {
+        startNextSegmentRead();
+      }
+      if (buffer.isEmpty()) {
+        if (future == null || lastFillComplete)
+          return false;
+        waitReadRowsFuture();
+      }
+      currentRow = FlatRowConverter.convert(buffer.remove());
+      return currentRow != null;
+    }
+
+    private SettableFuture<List<FlatRow>> createFuture() {
+      SettableFuture<List<FlatRow>> f = SettableFuture.create();
+
+      AtomicReference<ScanHandler> atomic = new AtomicReference<>();
+      ScanHandler handler;
+
+      handler = session.getDataClient().readFlatRows(buildReadRowsRequest(),
+          new StreamObserver<FlatRow>() {
+            List<FlatRow> rows = new ArrayList<>();
+            long currentByteSize = 0;
+            @Override
+            public void onNext(FlatRow flatRow) {
+              rows.add(flatRow);
+              currentByteSize += flatRow.getRowKey().size() + flatRow.getCells().stream()
+                  .mapToLong(c -> c.getQualifier().size() + c.getValue().size()).sum();
+              if (currentByteSize > bufferSizeLimit) {
+                atomic.get().cancel();
+                return;
+              }
+            }
+
+            @Override
+            public void onError(Throwable e) {
+              if (e instanceof CancellationException) {
+                f.set(rows);
+                return;
+              }
+              f.setException(e);
+            }
+
+            @Override
+            public void onCompleted() {
+              if (rows.size() < miniBatchLimit) {
+                lastFillComplete = true;
+              }
+              f.set(rows);
+              return;
+            }
+          });
+      atomic.set(handler);
+      return f;
+    }
+
+    private ReadRowsRequest buildReadRowsRequest() {
+      ReadRowsRequest.Builder request =
+          ReadRowsRequest.newBuilder()
+              .setRows(rowSet)
+              .setRowsLimit(miniBatchLimit)
+              .setTableName(tableName);
+      if (source.getRowFilter() != null) {
+        request.setFilter(source.getRowFilter());
+      }
+      return request.build();
+    }
+
+    private void startNextSegmentRead() {
```

Comment:
I think it would be better to inline this method and rename createFuture to startNextSegmentRead

Reply:
Done.